### PR TITLE
esp-idf: Fix build against latest esp-idf and for esp32-p4

### PR DESCRIPTION
**Pull request by tronical**

### CODE REVIEW

File: api/cpp/Cargo.toml [hunk 1]
```diff
@@ -57,7 +57,8 @@ raw-window-handle = { version = "0.5", optional = true }
 # Enable image-rs' default features to make all image formats to C++ users
 image = { version = "0.24.0", optional = true }
 
-esp-backtrace = { version = "0.9.0", features = ["panic-handler", "print-uart"], optional = true }
+esp-backtrace = { version = "0.11.0", features = ["panic-handler", "println"], optional = true }
+esp-println = { version = "0.9.0", default-features = false, features = ["uart"] }
```
Comment (ogoffart) [inline api/cpp/Cargo.toml:61]:
```suggestion
esp-println = { version = "0.9.0", default-features = false, features = ["uart"], optional = true }
```

Comment (tronical) [reply api/cpp/Cargo.toml]: Thanks!

File: api/cpp/esp-idf/slint/src/slint-esp.cpp [hunk 2]
```diff
@@ -230,12 +230,14 @@ void EspPlatform::run_event_loop()
                         std::swap(buffer1, buffer2.value());
                     } else {
                         for (int y = o.y; y < o.y + s.height; y++) {
+#if BSP_LCD_COLOR_SPACE == ESP_LCD_COLOR_SPACE_BGR
```
Comment (ogoffart) [inline api/cpp/esp-idf/slint/src/slint-esp.cpp:233]: We'll have to check if this still works on the S3 box  (that this is properly defined at this location.
(I'm not entirely sure how this gets defined if we don't include any bsp header)

File: api/cpp/esp-idf/slint/src/slint-esp.cpp [hunk 3]
```diff
@@ -6,7 +6,9 @@
 #include "slint-esp.h"
 #include "slint-platform.h"
 #include "esp_lcd_panel_ops.h"
-#include "esp_lcd_panel_rgb.h"
+#if SOC_LCD_RGB_SUPPORTED && ESP_IDF_VERSION_MAJOR >= 5
```
Comment (ogoffart) [inline api/cpp/esp-idf/slint/src/slint-esp.cpp:9]: SOC_LCD_RGB_SUPPORTED is defined indirectly from esp_lcd_panel_rgb.h via "soc/soc_caps.h". I'm not sure if the other header we include includes this.

File: api/cpp/Cargo.toml [hunk 4]
```diff
@@ -57,7 +57,8 @@ raw-window-handle = { version = "0.5", optional = true }
 # Enable image-rs' default features to make all image formats to C++ users
 image = { version = "0.24.0", optional = true }
 
-esp-backtrace = { version = "0.9.0", features = ["panic-handler", "print-uart"], optional = true }
+esp-backtrace = { version = "0.11.0", features = ["panic-handler", "println"], optional = true }
+esp-println = { version = "0.9.0", default-features = false, features = ["uart"], optional = true }
```
Comment (ogoffart) [inline api/cpp/Cargo.toml:61]: If I understand correctly, you depends on esp-println only to enable the uart feature.

I wonder if this works:

```suggestion
esp-backtrace = { version = "0.11.0", features = ["panic-handler", "println", "esp-println/uart"], optional = true }
```

(untested)

File: api/cpp/esp-idf/slint/src/slint-esp.cpp (hunk 2 above)
Comment (tronical) [reply api/cpp/esp-idf/slint/src/slint-esp.cpp:233]: I was curious about this, too, and added an #error inside this #ifdef and did builds for s3-usb-otb/s2-kaluga and s3-box, and it triggered for all of them. I'll double check to see where exactly it comes from.

File: api/cpp/Cargo.toml (hunk 4 above)
Comment (tronical) [reply api/cpp/Cargo.toml:61]: Unfortunately that doesn't work :(

```
  feature `esp-println/uart` in dependency `esp-backtrace` is not allowed to contain slashes
  If you want to enable features of a transitive dependency, the direct dependency needs to re-export those features from the `[features]` table.
```

File: api/cpp/esp-idf/slint/src/slint-esp.cpp (hunk 3 above)
Comment (ogoffart) [reply api/cpp/esp-idf/slint/src/slint-esp.cpp:9]: Maybe with __has_include ?

File: api/cpp/esp-idf/slint/src/slint-esp.cpp (hunk 2 above)
Comment (ogoffart) [reply api/cpp/esp-idf/slint/src/slint-esp.cpp:233]: maybe this is equivalent to 0 == 0  if none are defined?

Comment (ogoffart) [reply api/cpp/esp-idf/slint/src/slint-esp.cpp:233]: ESP_LCD_COLOR_SPACE_BGR is an enum value:
https://github.com/espressif/esp-idf/blob/60a2bf6a68867fd468a250966230a067f8d97ea2/components/esp_lcd/include/esp_lcd_types.h#L50
So i think this should somehow be a runtime check and not a compile time check.
btw, i think i looked for a way back in the days to check that but couldn't find a proper way. I think you'll need to pass an option to the init function of something like that.

**Comment by tronical** [comment]
Simplified down to just the two changes needed for this particular board.

File: api/cpp/esp-idf/slint/src/slint-esp.cpp (hunk 3 above)
Comment (tronical) [reply api/cpp/esp-idf/slint/src/slint-esp.cpp:10]: Good idea, can't harm :)

**Review by ogoffart — APPROVED** [review]

File: api/cpp/esp-idf/slint/src/slint-esp.cpp (hunk 3 above)
Comment (tronical) [reply api/cpp/esp-idf/slint/src/slint-esp.cpp:12]: Ohh, or do you mean for `esp_lcd_panel_rgb.h` and then make everything else conditional to that?

**Pull request merged**